pkg/deprecation/deprecation.go

```diff
@@ -0,0 +1,46 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package deprecation
+
+type Warning string
+
+const (
+	// Prefix is a standard prefix for all Warnings, used for filtering plugin Exports
+	Prefix = "io.containerd.deprecation/"
+	// PullSchema1Image is a warning for the use of schema 1 images
+	PullSchema1Image Warning = Prefix + "pull-schema-1-image"
+	// GoPluginLibrary is a warning for the use of dynamic library Go plugins
+	GoPluginLibrary Warning = Prefix + "go-plugin-library"
+)
+
+var messages = map[Warning]string{
+	PullSchema1Image: "Schema 1 images are deprecated since containerd v1.7 and removed in containerd v2.0. " +
```

Comment:
So the message for a specific deprecation is basically hard-coded, and the plugins (i.e. clients of the warning service) cannot put free-text (e.g. we might want to add image name to the message) into the message.  Should we consider loosening it up by allowing plugins to call `Emit(ID, message, timestamp)` , and the warning service will use `ID` to decide whether to accept or ignore the deprecation report?

Reply:
It's intentionally limited for a couple reasons:

1. A free-form message would need limits to ensure that there's no unbounded memory growth in the warning service and no unbounded growth in the gRPC payload as part of the introspection service's `Server` rpc response.
2. For something like schema 1 images, there may be many schema 1 images that trigger the warning, but we only want one entry returned.  The hard-coded message can direct clients to discover more details elsewhere if desired.